### PR TITLE
[SPARK-31110][DOCS][SQL] refine sql doc for SELECT

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

A few improvements to the sql ref SELECT doc:
1. correct the syntax of SELECT query
2. correct the default of null sort order
3. correct the GROUP BY syntax
4. several minor fixes

### Why are the changes needed?

refine document

### Does this PR introduce any user-facing change?

N/A

### How was this patch tested?

N/A

### CODE REVIEW

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 1]
```diff
@@ -9,9 +9,9 @@ license: |
   The ASF licenses this file to You under the Apache License, Version 2.0
   (the "License"); you may not use this file except in compliance with
   the License.  You may obtain a copy of the License at
- 
+
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: My editor removes trailing spaces automatically

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 2]
```diff
@@ -9,33 +9,33 @@ license: |
...
-the group of rows based on one or more specified aggregate functions. Spark also supports advanced aggregations to do multiple 
+The <code>GROUP BY</code> clause is used to group the rows based on a set of specified grouping expressions and compute aggregations on
+the group of rows based on one or more specified aggregate functions. Spark also supports advanced aggregations to do multiple
 aggregations for the same input record set via `GROUPING SETS`, `CUBE`, `ROLLUP` clauses.
 
 ### Syntax
 {% highlight sql %}
 GROUP BY [ GROUPING SETS grouping_sets ] group_expression [ , group_expression [ , ... ] ]
-    [ ( WITH ROLLUP | WITH CUBE | GROUPING SETS grouping_sets ) ) ]
+    [ ( WITH ROLLUP | WITH CUBE | GROUPING SETS grouping_sets ) ]
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-groupby.md]: remove the extra `)`

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 3]
```diff
@@ -55,25 +55,26 @@ GROUP BY [ GROUPING SETS grouping_sets ] group_expression [ , group_expression [
   </dd>
   <dt><code><em>ROLLUP</em></code></dt>
   <dd>
-    Specifies multiple levels of aggregations in a single statement. This clause is used to compute aggregations 
-    based on multiple grouping sets. <code>ROLLUP</code> is shorthand for <code>GROUPING SETS</code>. For example,
-    GROUP BY warehouse, product  WITH ROLLUP is equivalent to GROUP BY <code>warehouse, product</code> GROUPING SETS
-    <code> ((warehouse, product), (warehouse), ())</code>.
+    Specifies multiple levels of aggregations in a single statement. This clause is used to compute aggregations
+    based on multiple grouping sets. <code>ROLLUP</code> is a shorthand for <code>GROUPING SETS</code>. For example,
+    <code>GROUP BY warehouse, product WITH ROLLUP</code> is equivalent to <code>GROUP BY GROUPING SETS
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-groupby.md]: GROUP BY GROUPING SETS ... is the same as GROUP BY a, b GROUPING SETS but shorter.

File: docs/sql-ref-syntax-qry-select-orderby.md [hunk 4]
```diff
@@ -47,20 +47,18 @@ ORDER BY { expression [ sort_direction | nulls_sort_oder ] [ , ... ] }
...
   <dt><code><em>nulls_sort_order</em></code></dt>
   <dd>
-    Optionally specifies whether NULL values are returned before/after non-NULL values, based on the 
-    sort direction. In Spark, NULL values are considered to be lower than any non-NULL values by default.
-    Therefore the ordering of NULL values depend on the sort direction. If <code>null_sort_order</code> is
-    not specified, then NULLs sort first if sort order is <code>ASC</code> and NULLS sort last if 
-    sort order is <code>DESC</code>.<br><br>
+    Optionally specifies whether NULL values are returned before/after non-NULL values. If
+    <code>null_sort_order</code> is not specified, then NULLs sort first if sort order is
+    <code>ASC</code> and NULLS sort last if sort order is <code>DESC</code>.<br><br>
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-orderby.md]: We don't need to mention `In Spark, NULL values are considered to be lower than any non-NULL values`, as it's not related to the user-facing behavior.

File: docs/sql-ref-syntax-qry-select-orderby.md [hunk 5]
```diff
@@ -47,20 +47,18 @@ ORDER BY { expression [ sort_direction | nulls_sort_oder ] [ , ... ] }
...
     <ol>
-      <li> If <code>NULLS FIRST</code> (the default) is specified, then NULL values are returned first 
+      <li> If <code>NULLS FIRST</code> is specified, then NULL values are returned first
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-orderby.md]: the default value depends on the sort order.

File: docs/sql-ref-syntax-qry-select.md [hunk 6]
```diff
@@ -1,57 +1,61 @@
...
+used to retrieve result sets from one or more tables. The following section
 describes the overall query syntax and the sub-sections cover different constructs
-of a query along with examples. 
+of a query along with examples.
 
 ### Syntax
 {% highlight sql %}
 [ WITH with_query [ , ... ] ]
+select_statement [ { UNION | INTERSECT | EXCEPT } select_statement, ... ]
+[ ORDER BY { expression [ ASC | DESC ] [ NULLS { FIRST | LAST } ] [ , ...] } ]
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select.md]: according to our g4 file, a query can have multiple SELECT but only one ORDER BY/SORT BY/...

**Comment by cloud-fan** [comment]
@dilipbiswal @dongjoon-hyun @viirya @maropu

File: docs/sql-ref-syntax-qry-select.md [hunk 7]
```diff
@@ -1,57 +1,61 @@
...
-used to retrieve result sets from one or more tables. The following section 
+used to retrieve result sets from one or more tables. The following section
 describes the overall query syntax and the sub-sections cover different constructs
-of a query along with examples. 
+of a query along with examples.
 
 ### Syntax
 {% highlight sql %}
 [ WITH with_query [ , ... ] ]
+select_statement [ { UNION | INTERSECT | EXCEPT } select_statement, ... ]
```
Comment (dongjoon-hyun) [inline docs/sql-ref-syntax-qry-select.md]: `[ ALL | DISTINCT ]` seems to be missed while transition.

File: docs/sql-ref-syntax-qry-select.md [hunk 8]
```diff
@@ -1,57 +1,61 @@
...
+[ ORDER BY { expression [ ASC | DESC ] [ NULLS { FIRST | LAST } ] [ , ...] } ]
+[ SORT  BY { expression [ ASC | DESC ] [ NULLS { FIRST | LAST } ] [ , ...] } ]
```
Comment (dongjoon-hyun) [inline docs/sql-ref-syntax-qry-select.md]: `SORT  BY` -> `SORT BY` (double space).

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with two minor comments.)
Thank you, @cloud-fan .

File: docs/sql-ref-syntax-qry-select-limit.md [hunk 9]
```diff
@@ -88,6 +88,11 @@ SELECT name, age FROM person ORDER BY name LIMIT length('SPARK')
   | Mike A| 25|
   |Shone S| 16|
   +-------+---+
+
+-- A non-foldable expression as an input to limit is not allowed.
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-limit.md]: nit: `limit` -> `LIMIT`?

**Review by maropu — APPROVED** [review]
The change can make the docs cleaner! LGTM except the existing comments.

File: docs/sql-ref-syntax-qry-select.md (hunk 7 above)
Comment (cloud-fan) [reply docs/sql-ref-syntax-qry-select.md]: It's defined in `select_statement`: https://github.com/apache/spark/pull/27866/files#diff-bc52347d9c1373190bce4389d3066ad5R40

File: docs/sql-ref-syntax-qry-select-groupby.md [hunk 10]
```diff
@@ -9,42 +9,43 @@ license: |
...
 ---
-The <code>GROUP BY</code> clause is used to group the rows based on a set of specified grouping expressions and compute aggregations on 
-the group of rows based on one or more specified aggregate functions. Spark also supports advanced aggregations to do multiple 
+The <code>GROUP BY</code> clause is used to group the rows based on a set of specified grouping expressions and compute aggregations on
+the group of rows based on one or more specified aggregate functions. Spark also supports advanced aggregations to do multiple
 aggregations for the same input record set via `GROUPING SETS`, `CUBE`, `ROLLUP` clauses.
 
 ### Syntax
 {% highlight sql %}
-GROUP BY [ GROUPING SETS grouping_sets ] group_expression [ , group_expression [ , ... ] ]
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-qry-select-groupby.md]: This was wrong.

Comment (cloud-fan) [reply docs/sql-ref-syntax-qry-select-groupby.md]: @viirya can you help review this file? thanks!

File: docs/sql-ref-syntax-qry-select.md (hunk 7 above)
Comment (dongjoon-hyun) [reply docs/sql-ref-syntax-qry-select.md]: Hi, @cloud-fan . They are not the same. I mean `UNION ALL`, not `SELECT ALL`.
```
scala> sql("select 1 union all (select all 1)").show
+---+
|  1|
+---+
|  1|
|  1|
+---+
```

**Comment by dongjoon-hyun** [comment]
@cloud-fan . Please don't forget the following
- https://github.com/apache/spark/pull/27866#discussion_r391115170

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master/3.0. Thank you all!

File: docs/sql-ref-syntax-qry-select-groupby.md (hunk 10 above)
Comment (viirya) [reply docs/sql-ref-syntax-qry-select-groupby.md]: Sorry for late. This LGTM.

**Pull request closed without merging**